File: tradex/agent/tools/registry.py
"""工具系统基础设施：注册表、数据结构、钩子类型和执行引擎。"""
from __future__ import annotations

import json
from dataclasses import dataclass
from inspect import isawaitable
from typing import Any, Awaitable, Callable

ToolHandler = Callable[..., Awaitable[str]]


@dataclass(frozen=True)
class ToolDefinition:
    """一个可被 LLM 调用的工具定义。"""

    name: str
    description: str
    parameters: dict[str, Any]
    handler: ToolHandler


@dataclass
class ToolCall:
    """模型返回的一次工具调用。"""

    id: str
    name: str
    arguments: dict[str, Any]


@dataclass
class ToolResult:
    """工具执行结果。"""

    call_id: str
    name: str
    output: str
    error: bool = False


BeforeToolHook = Callable[[ToolCall, ToolDefinition], ToolCall | None | Awaitable[ToolCall | None]]
AfterToolHook = Callable[
    [ToolCall, ToolResult, ToolDefinition],
    ToolResult | None | Awaitable[ToolResult | None],
]


class ToolRegistry:
    """管理所有可用工具的注册表。"""

    def __init__(
        self,
        *,
        before_tool_hooks: tuple[BeforeToolHook, ...] = tuple(),
        after_tool_hooks: tuple[AfterToolHook, ...] = tuple(),
    ) -> None:
        """初始化工具注册表，可选配置调用前后的钩子。"""
        self._tools: dict[str, ToolDefinition] = {}
        self._before_tool_hooks = before_tool_hooks
        self._after_tool_hooks = after_tool_hooks

    def extend_hooks(
        self,
        *,
        before_tool_hooks: tuple[BeforeToolHook, ...] = tuple(),
        after_tool_hooks: tuple[AfterToolHook, ...] = tuple(),
    ) -> None:
        """追加 runtime 级工具钩子，用于审计、权限或参数改写。"""
        self._before_tool_hooks = (*self._before_tool_hooks, *before_tool_hooks)
        self._after_tool_hooks = (*self._after_tool_hooks, *after_tool_hooks)

    def register(self, tool: ToolDefinition) -> None:
        """注册一个工具定义到注册表中。"""
        self._tools[tool.name] = tool

    def get(self, name: str) -> ToolDefinition | None:
        """按名称查找并返回工具定义。"""
        return self._tools.get(name)

    def list_tools(self) -> list[ToolDefinition]:
        """返回所有已注册工具的列表。"""
        return list(self._tools.values())

    def openai_tool_schemas(self) -> list[dict[str, Any]]:
        """生成 OpenAI function calling 格式的 tool schema 列表。"""
        schemas: list[dict[str, Any]] = []
        for tool in self._tools.values():
            schemas.append({
                "type": "function",
                "function": {
                    "name": tool.name,
                    "description": tool.description,
                    "parameters": tool.parameters,
                },
            })
        return schemas

    def codex_tool_schemas(self) -> list[dict[str, Any]]:
        """生成 Codex Responses API 格式的 tool schema 列表。"""
        schemas: list[dict[str, Any]] = []
        for tool in self._tools.values():
            schemas.append({
                "type": "function",
                "name": tool.name,
                "description": tool.description,
                "parameters": tool.parameters,
            })
        return schemas

    async def execute(self, call: ToolCall) -> ToolResult:
        """执行一次工具调用并返回结果。"""
        tool = self._tools.get(call.name)
        if tool is None:
            return ToolResult(
                call_id=call.id,
                name=call.name,
                output=f"Unknown tool: {call.name}",
                error=True,
            )
        effective_call = call
        try:
            for hook in self._before_tool_hooks:
                replacement = await _maybe_await(hook(effective_call, tool))
                if isinstance(replacement, ToolCall):
                    effective_call = replacement
            output = await tool.handler(**effective_call.arguments)
            result = ToolResult(call_id=effective_call.id, name=effective_call.name, output=output)
        except Exception as exc:
            result = ToolResult(
                call_id=effective_call.id,
                name=effective_call.name,
                output=str(exc) or exc.__class__.__name__,
                error=True,
            )
        for hook in self._after_tool_hooks:
            replacement = await _maybe_await(hook(effective_call, result, tool))
            if isinstance(replacement, ToolResult):
                result = replacement
        return result


def _json_output(data: Any) -> str:
    """将数据序列化为紧凑的 JSON 字符串。"""
    return json.dumps(data, ensure_ascii=False, separators=(",", ":"))


async def _maybe_await(value: Any) -> Any:
    """若值为可等待对象则 await，否则直接返回。"""
    if isawaitable(value):
        return await value
    return value


def merge_registries(*registries: ToolRegistry) -> ToolRegistry:
    """合并多个工具注册表到一个新注册表。"""
    merged = ToolRegistry()
    for registry in registries:
        for tool in registry.list_tools():
            merged.register(tool)
    return merged

File: tradex/agent/tools/registry.ts
/**
 * tools/registry.ts — Tool definition + registry.
 *
 * Each tool has a single `execute` method. The canonical result shape is
 * { content, details?, terminate? } where `content` is an array of
 * TextContent | ImageContent blocks. Errors are signaled by throwing.
 *
 * For pragmatic back-compat with tradex's existing 30+ tools, `execute` may
 * also return:
 *   • a plain string (auto-wrapped to [{ type: "text", text }]); or
 *   • a bare ContentBlock[] (treated as { content, details: undefined }).
 *
 * The tool-adapter normalizes all three shapes before handing them to the
 * agent loop. New tools should prefer the canonical result shape.
 */

import type { AgentRuntimeId } from "../runtime/types.js";

export interface TextContent {
  type: "text";
  text: string;
}

export interface ImageContent {
  type: "image";
  data: string;
  mimeType: string;
}

export type ToolExecutionMode = "sequential" | "parallel";
export type ToolProgressCallback<TDetails = unknown> = (update: ToolHandlerResult<TDetails>) => void;

/** Single text or image block returned to the model. */
export type ContentBlock = TextContent | ImageContent;

/**
 * Canonical tool execution result.
 */
export interface ToolHandlerResult<TDetails = unknown> {
  /** Content blocks returned to the model. */
  content: ContentBlock[];
  /** Optional structured details for logs / UI rendering (not sent to model). */
  details?: TDetails;
  /**
   * Hint that the agent loop should stop after this tool batch.
   * Termination only occurs when every finalized tool result in the batch
   * sets this to true.
   */
  terminate?: boolean;
}

/**
 * Permitted return values for `ToolDefinition.execute`.
 *
 * The first two forms are convenience sugar for tools that only ever return
 * text or only ever return content blocks. The third form is the canonical
 * result that supports `details` and `terminate`.
 */
export type ToolReturnValue<TDetails = unknown> =
  | string
  | ContentBlock[]
  | ToolHandlerResult<TDetails>;

/**
 * Tool execute signature (without `toolCallId` first argument).
 */
export type ToolExecuteFn<TDetails = unknown> = (
  args: Record<string, unknown>,
  signal?: AbortSignal,
  onUpdate?: ToolProgressCallback<TDetails>,
) => Promise<ToolReturnValue<TDetails>> | ToolReturnValue<TDetails>;

export interface ToolPolicy {
  access: "read" | "write";
  domain: "market" | "news" | "browser" | "filesystem" | "trading" | "external" | "other";
  runtimeExposure: readonly AgentRuntimeId[];
}

export interface ToolDefinition<TDetails = unknown> {
  name: string;
  description: string;
  parameters: Record<string, unknown>;
  /**
   * Execute the tool call. Throw on failure — unhandled errors are caught by
   * the registry / adapter and surfaced as a tool error to the model.
   */
  execute: ToolExecuteFn<TDetails>;
  /** Per-tool execution mode override. */
  executionMode?: ToolExecutionMode;
  /** Missing policy is deliberately Pi-only. */
  policy?: ToolPolicy;
}

export interface ToolCall {
  id: string;
  name: string;
  arguments: Record<string, unknown>;
}

export interface ToolResult {
  callId: string;
  name: string;
  output: string;
  error: boolean;
}

export type BeforeToolHook = (call: ToolCall, tool: ToolDefinition) => ToolCall | null | Promise<ToolCall | null>;
export type AfterToolHook = (call: ToolCall, result: ToolResult, tool: ToolDefinition) => ToolResult | null | Promise<ToolResult | null>;

/** Coerce any permitted ToolReturnValue into a canonical ToolHandlerResult. */
export function normalizeToolReturn<TDetails = unknown>(
  value: ToolReturnValue<TDetails>,
): ToolHandlerResult<TDetails> {
  if (typeof value === "string") {
    return { content: [{ type: "text", text: value }] };
  }
  if (Array.isArray(value)) {
    return { content: value };
  }
  return value;
}

/** Flatten content blocks into the plain-text representation used by ToolResult.output. */
export function contentToText(content: ContentBlock[]): string {
  return content
    .filter((c): c is TextContent => c.type === "text")
    .map((c) => c.text)
    .join("\n");
}

export class ToolRegistry {
  private readonly tools = new Map<string, ToolDefinition>();
  private beforeToolHooks: BeforeToolHook[];
  private afterToolHooks: AfterToolHook[];

  constructor(input: { beforeToolHooks?: BeforeToolHook[]; afterToolHooks?: AfterToolHook[] } = {}) {
    this.beforeToolHooks = input.beforeToolHooks ?? [];
    this.afterToolHooks = input.afterToolHooks ?? [];
  }

  extendHooks(input: { beforeToolHooks?: BeforeToolHook[]; afterToolHooks?: AfterToolHook[] }): void {
    this.beforeToolHooks.push(...(input.beforeToolHooks ?? []));
    this.afterToolHooks.push(...(input.afterToolHooks ?? []));
  }

  register(tool: ToolDefinition): void {
    this.tools.set(tool.name, tool);
  }

  get(name: string): ToolDefinition | null {
    return this.tools.get(name) ?? null;
  }

  setPolicy(name: string, policy: ToolPolicy): void {
    const tool = this.tools.get(name);
    if (!tool) throw new Error(`Unknown tool: ${name}`);
    tool.policy = policy;
  }

  listTools(): ToolDefinition[] {
    return [...this.tools.values()];
  }

  listToolsForRuntime(runtime: AgentRuntimeId, access: "read" | "write" = "read"): ToolDefinition[] {
    return this.listTools().filter((tool) => {
      const policy = tool.policy ?? { access: "read" as const, domain: "other" as const, runtimeExposure: ["pi"] as const };
      return policy.runtimeExposure.includes(runtime) && (access === "write" || policy.access === "read");
    });
  }

  openaiToolSchemas(): Array<Record<string, unknown>> {
    return this.listTools().map((tool) => ({
      type: "function",
      function: {
        name: tool.name,
        description: tool.description,
        parameters: tool.parameters,
      },
    }));
  }

  codexToolSchemas(): Array<Record<string, unknown>> {
    return this.listTools().map((tool) => ({
      type: "function",
      name: tool.name,
      description: tool.description,
      parameters: tool.parameters,
    }));
  }

  async execute(call: ToolCall): Promise<ToolResult> {
    const tool = this.tools.get(call.name);
    if (!tool) return { callId: call.id, name: call.name, output: `Unknown tool: ${call.name}`, error: true };
    let effectiveCall = call;
    try {
      for (const hook of this.beforeToolHooks) {
        const replacement = await hook(effectiveCall, tool);
        if (replacement) effectiveCall = replacement;
      }
      const raw = await tool.execute(effectiveCall.arguments);
      const normalized = normalizeToolReturn(raw);
      const output = contentToText(normalized.content);
      let result: ToolResult = { callId: effectiveCall.id, name: effectiveCall.name, output, error: false };
      for (const hook of this.afterToolHooks) {
        const replacement = await hook(effectiveCall, result, tool);
        if (replacement) result = replacement;
      }
      return result;
    } catch (error) {
      return {
        callId: effectiveCall.id,
        name: effectiveCall.name,
        output: error instanceof Error ? error.message : String(error),
        error: true,
      };
    }
  }
}

export function jsonOutput(data: unknown): string {
  return JSON.stringify(data);
}

export function mergeRegistries(...registries: ToolRegistry[]): ToolRegistry {
  const merged = new ToolRegistry();
  for (const registry of registries) {
    for (const tool of registry.listTools()) merged.register(tool);
  }
  return merged;
}
